sbom/spdx/encoder.go

```diff
@@ -0,0 +1,377 @@
+	"github.com/spdx/tools-golang/spdx/v2/v2_3"
+)
+
+type Version string
+
+const (
+	V2_3 Version = "v2.3"
+)
+
+type Format string
+
+const JSON Format = "json"
+
+type Creator struct {
+	Creator string
+	// In accordance to the SPDX v2 spec, CreatorType should be one of "Person", "Organization", or "Tool"
+	CreatorType string
+}
+
+var _ sbom.Encoder = (*Encoder)(nil)
+
+type Encoder struct {
+	Version           Version
+	Format            Format
+	Creators          []Creator
+	DocumentName      string
+	DocumentNamespace string
+	DocumentComment   string
+}
+
+// Encode encodes a claircore IndexReport to an io.Reader.
+// We first convert the IndexReport to an SPDX doc of the latest version, then
+// convert that doc to the specified version. We assume there's no data munging
+// going from latest to the specified version.
+func (e *Encoder) Encode(ctx context.Context, ir *claircore.IndexReport) (io.Reader, error) {
+	spdx, err := e.parseIndexReport(ctx, ir)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(blugo): support SPDX versions before 2.3
+	var tmpConverterDoc common.AnyDocument
+	switch e.Version {
+	case V2_3:
+		// parseIndexReport currently returns a v2_3.Document so do nothing
+		tmpConverterDoc = spdx
+	default:
+		return nil, fmt.Errorf("unknown SPDX version: %v", e.Version)
+	}
+
+	switch e.Format {
+	case JSON:
+		buf := &bytes.Buffer{}
+		if err := spdxjson.Write(tmpConverterDoc, buf); err != nil {
+			return nil, err
+		}
+		return buf, nil
+	}
+
+	return nil, fmt.Errorf("unknown requested format: %v", e.Format)
+}
+
+func (e *Encoder) parseIndexReport(ctx context.Context, ir *claircore.IndexReport) (*v2_3.Document, error) {
+	creatorInfo := e.Creators
+	spdxCreators := make([]v2common.Creator, len(creatorInfo))
+	for i, creator := range creatorInfo {
+		spdxCreators[i].Creator = creator.Creator
+		spdxCreators[i].CreatorType = creator.CreatorType
+	}
+
+	// Initial metadata
+	out := &v2_3.Document{
+		SPDXVersion:       v2_3.Version,
+		DataLicense:       v2_3.DataLicense,
+		SPDXIdentifier:    "DOCUMENT",
+		DocumentName:      e.DocumentName,
+		DocumentNamespace: e.DocumentNamespace,
+		CreationInfo: &v2_3.CreationInfo{
+			Creators: spdxCreators,
+			Created:  time.Now().Format("2006-01-02T15:04:05Z"),
+		},
+		DocumentComment: e.DocumentComment,
+	}
+
+	pkgMap := map[int]*v2_3.Package{}
+	var pkgIds []int
+	distMap := map[int]*v2_3.Package{}
+	var distIds []int
+	repoMap := map[int]*v2_3.Package{}
+	var repoIds []int
+	pkgRels := map[int][]*v2_3.Relationship{}
+	for _, r := range ir.IndexRecords() {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		rPkgId, err := strconv.Atoi(r.Package.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		pkg, ok := pkgMap[rPkgId]
+		// Record the package if we haven't seen it yet.
+		if !ok {
+			pkgDB := ""
+			for _, env := range ir.Environments[r.Package.ID] {
+				if env.PackageDB != "" {
+					pkgDB = env.PackageDB
```

Comment:
mentioned offline, but this has a prefix, which we may want to consider removing and putting elsewhere

Reply:
I'm open to suggestions. This seemed like a _fine_ alternative in lieu of pURLs.